src/Authentication/Authentication/Cmdlets/InvokeGraphRequest.cs

```diff
@@ -0,0 +1,965 @@
+                var encodedKey = WebUtility.UrlEncode(key);
+                var encodedValue = string.Empty;
+                if (value != null) encodedValue = WebUtility.UrlEncode(value.ToString());
+
+                bodyBuilder.AppendFormat("{0}={1}", encodedKey, encodedValue);
+            }
+
+            return bodyBuilder.ToString();
+        }
+
+        private HttpRequestMessage GetRequest(Uri uri)
+        {
+            var requestUri = PrepareUri(uri);
+            var httpMethod = GetHttpMethod(Method);
+            // create the base WebRequest object
+            var request = new HttpRequestMessage(httpMethod, requestUri);
+
+            // pull in session data
+            if (GraphRequestSession.Headers.Count > 0)
+            {
+                GraphRequestSession.ContentHeaders.Clear();
+                foreach (var entry in GraphRequestSession.Headers)
+                {
+                    if (HttpKnownHeaderNames.ContentHeaders.Contains(entry.Key))
+                    {
+                        GraphRequestSession.ContentHeaders.Add(entry.Key, entry.Value);
+                    }
+                    else
+                    {
+                        if (SkipHeaderValidation)
+                        {
+                            request.Headers.TryAddWithoutValidation(entry.Key, entry.Value);
+                        }
+                        else
+                        {
+                            request.Headers.Add(entry.Key, entry.Value);
+                        }
+                    }
+                }
+            }
+
+            // Set 'Transfer-Encoding: chunked' if 'Transfer-Encoding' is specified
+            if (GraphRequestSession.Headers.ContainsKey(HttpKnownHeaderNames.TransferEncoding))
+            {
+                request.Headers.TransferEncodingChunked = true;
+            }
+
+            // Set 'User-Agent' if WebSession.Headers doesn't already contain it
+            if (GraphRequestSession.Headers.TryGetValue(HttpKnownHeaderNames.UserAgent, out var userAgent))
+            {
+                GraphRequestSession.UserAgent = userAgent;
+            }
+            else
+            {
+                if (SkipHeaderValidation)
+                {
+                    request.Headers.TryAddWithoutValidation(HttpKnownHeaderNames.UserAgent,
+                        GraphRequestSession.UserAgent);
+                }
+                else
+                {
+                    request.Headers.Add(HttpKnownHeaderNames.UserAgent, GraphRequestSession.UserAgent);
+                }
+            }
+
+            return request;
+        }
+
+        private Uri PrepareUri(Uri uri)
+        {
+            // before creating the web request,
+            // preprocess Body if content is a dictionary and method is GET (set as query)
+            LanguagePrimitives.TryConvertTo(Body, out IDictionary bodyAsDictionary);
+            if (bodyAsDictionary != null && Method == GraphRequestMethod.GET)
```

Comment:
PERF: if you change these two lines into:
`if (Method == GraphRequestMethod.GET && LanguagePrimitives.TryConvertTo(Body, out IDictionary bodyAsDictionary))`

you will avoid converting the body into dictionary if the method is not `GET`.